stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/channel/list/adapter/viewholder/ChannelItemSwipeListener.kt

```diff
@@ -0,0 +1,135 @@
+package io.getstream.chat.android.ui.channel.list.adapter.viewholder
+
+import android.content.Context
+import android.view.View
+import androidx.recyclerview.widget.RecyclerView
+import io.getstream.chat.android.ui.R
+import io.getstream.chat.android.ui.channel.list.ChannelListView
+import io.getstream.chat.android.ui.channel.list.ScrollPauseLinearLayoutManager
+import io.getstream.chat.android.ui.utils.extensions.getDimension
+import io.getstream.chat.android.ui.utils.extensions.safeCast
+
+internal class ChannelItemSwipeListener(
+    context: Context,
+    val recyclerView: RecyclerView,
+    val layoutManager: ScrollPauseLinearLayoutManager
+) : ChannelListView.SwipeListener {
+
+    private val menuItemWidth = context.getDimension(R.dimen.stream_ui_channel_list_item_option_icon_width).toFloat()
+    private val optionsMenuWidth = menuItemWidth * ChannelItemViewHolder.OPTIONS_COUNT
+    private val openValue = -optionsMenuWidth
+    private val closedValue = 0f
+    private val swipeRange = openValue..closedValue
+    private val swipeStateByPosition = mutableMapOf<Int, SwipeState>()
+    var multiSwipeEnabled: Boolean = false
+
+    // Represents the internal swipe state of the ViewHolder
+    sealed class SwipeState {
+        internal object Open : SwipeState()
+        internal object Closed : SwipeState()
+    }
+
+    override fun onSwipeStarted(viewHolder: RecyclerView.ViewHolder, adapterPosition: Int, x: Float, y: Float) {
+        // pause scrolling
+        layoutManager.verticalScrollEnabled = false
+    }
+
+    override fun onSwipeChanged(viewHolder: RecyclerView.ViewHolder, adapterPosition: Int, dX: Float) {
+        viewHolder.safeCast<ChannelItemViewHolder>()?.let { channelViewHolder ->
```

Comment:
This `safeCast` feels very C++ to me, and I don't know what's the benefit on top of the actual safe cast syntax. But it doesn't bother me too much either.

Reply:
I just like it when chaining for ease and aesthetics